File: sdk/php/src/Internal/Utils/SubInfo.php
<?php

namespace KuCoin\UniversalSDK\Internal\Utils;

use Exception;
use KuCoin\UniversalSDK\Internal\Interfaces\WebSocketMessageCallback;

class SubInfo
{
    const EMPTY_ARGS_STR = 'EMPTY_ARGS';

    /** @var string */
    public $prefix;

    /** @var string[] */
    public $args;

    /** @var WebSocketMessageCallback|null */
    public $callback;

    /**
     * @param string $prefix
     * @param string[] $args
     * @param WebSocketMessageCallback|null $callback
     */
    public function __construct(string $prefix, array $args = [], WebSocketMessageCallback $callback = null)
    {
        $this->prefix = $prefix;
        $this->args = $args;
        $this->callback = $callback;
    }

    /**
     * @return string
     */
    public function toId(): string
    {
        if (empty($this->args)) {
            $argsStr = self::EMPTY_ARGS_STR;
        } else {
            $sortedArgs = $this->args;
            sort($sortedArgs, SORT_STRING);
            $argsStr = implode(',', $sortedArgs);
        }

        return $this->prefix . '@@' . $argsStr;
    }

    /**
     * @param string $id
     * @param callable|null $callback
     * @return SubInfo
     * @throws Exception
     */
    public static function fromId(string $id, WebSocketMessageCallback $callback = null): SubInfo
    {
        $parts = explode('@@', $id, 2);
        if (count($parts) !== 2) {
            throw new Exception("SubInfo::fromId: invalid id format: $id");
        }

        $prefix = $parts[0];
        $args = ($parts[1] === self::EMPTY_ARGS_STR) ? [] : explode(',', $parts[1]);

        return new self($prefix, $args, $callback);
    }

    /**
     * @return string[]
     */
    public function topics(): array
    {
        if (empty($this->args)) {
            return [$this->prefix];
        }

        $topics = [];
        foreach ($this->args as $arg) {
            $topics[] = $this->prefix . ':' . $arg;
        }

        return $topics;
    }

    /**
     * @return string
     */
    public function subTopic(): string
    {
        if (empty($this->args)) {
            return $this->prefix;
        }

        return $this->prefix . ':' . implode(',', $this->args);
    }
}

File: sdk/java/src/main/java/com/kucoin/universal/sdk/internal/infra/SubInfo.java
package com.kucoin.universal.sdk.internal.infra;

import com.kucoin.universal.sdk.internal.interfaces.WebSocketMessageCallback;
import java.util.*;
import java.util.stream.Collectors;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@Getter
public final class SubInfo {

  private static final String SEP = "@@";
  private static final String EMPTY_ARGS_STR = "EMPTY_ARGS";

  private final String prefix;
  private final List<String> args;
  private final WebSocketMessageCallback callback;

  public String toId() {
    if (args == null || args.isEmpty()) {
      return prefix + SEP + EMPTY_ARGS_STR;
    }

    List<String> sorted = new ArrayList<>(args);
    Collections.sort(sorted);
    return prefix + SEP + String.join(",", sorted);
  }

  public static SubInfo fromId(String id) {
    String[] parts = id.split(SEP, -1);
    if (parts.length != 2) throw new IllegalArgumentException("invalid id: " + id);

    String prefix = parts[0];
    List<String> args;
    if (parts[1].equals(EMPTY_ARGS_STR)) {
      args = new LinkedList<>();
    } else {
      args = Arrays.asList(parts[1].split(","));
    }
    return new SubInfo(prefix, args, null);
  }

  public String subTopic() {
    return args == null || args.isEmpty() ? prefix : prefix + ":" + String.join(",", args);
  }

  public Set<String> topics() {
    if (args == null || args.isEmpty()) {
      return Collections.singleton(prefix);
    }
    return args.stream().map(a -> prefix + ":" + a).collect(Collectors.toSet());
  }
}
